File: test/test/math/TestMatrix.java
/**
 * 2018-01-25
 */
package test.math;

import cn.ancono.math.MathCalculator;
import cn.ancono.math.algebra.linearAlgebra.LinearEquationSolution;
import cn.ancono.math.algebra.linearAlgebra.Matrix;
import cn.ancono.math.algebra.linearAlgebra.MatrixSup;
import cn.ancono.math.algebra.linearAlgebra.Vector;
import cn.ancono.math.equation.SVPEquation;
import cn.ancono.math.geometry.analytic.planeAG.curve.ConicSection;
import cn.ancono.math.geometry.analytic.planeAG.curve.GeneralConicSection;
import cn.ancono.math.numberModels.Calculators;
import cn.ancono.math.numberModels.Fraction;
import cn.ancono.utilities.ArraySup;
import cn.ancono.utilities.structure.Pair;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

/**
 * @author liyicheng 2018-01-25 17:38
 *
 */
public class TestMatrix {

    /**
     *
     */
    public TestMatrix() {
    }

    MathCalculator<Long> mc = Calculators.getCalLong();
    MathCalculator<Double> mcd = Calculators.getCalDoubleDev();

    private <T> boolean isUpperTriangular(Matrix<T> m) {
        var mc = m.getMathCalculator();
        var n = Math.min(m.getRowCount(), m.getColumnCount());
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (!mc.isZero(m.getNumber(i, j))) {
                    return false;
                }
            }
        }
        return true;
    }

    //	@Test
    public void testEigenEquation() {
        Matrix<Long> mat = Matrix.of(new long[][]{{1, 0}, {0, 4}});
        SVPEquation<Long> equation = mat.charEquation(), expected = SVPEquation.quadratic(1L, -5L, 4L, mc);
        assertTrue("EigenEquation:", expected.valueEquals(equation));
        mat = Matrix.of(new long[][]{{1, 2}, {3, 4}});
        equation = mat.charEquation();
        expected = SVPEquation.quadratic(1L, -5L, -2L, mc);
        assertTrue("EigenEquation:", expected.valueEquals(equation));
        mat = Matrix.of(new long[][]{{1, 2, 3}, {3, 4, 5}, {4, 5, 6}});
        equation = mat.charEquation();
        expected = SVPEquation.valueOf(mc, 0L, -9L, -11L, 1L);
        assertTrue("EigenEquation:", expected.valueEquals(equation));
    }

//	@Test
	public void testSolveEquation() {
		int row = 8;
		int column = 10;
		for (int i = 0; i < 100; i++) {
            double[][] mat = new double[row][];
            for (int j = 0; j < row; j++) {
                mat[j] = ArraySup.ranDoubleArr(column);
            }
            Matrix<Double> matrix = Matrix.of(mat).mapTo(x -> x, mcd);
            LinearEquationSolution<Double> solution = MatrixSup.solveLinearEquation(matrix);
            if (solution.getSolutionSituation() != LinearEquationSolution.Situation.NO_SOLUTION) {
                Vector<Double> base = solution.getSpecialSolution();
                Vector<Double>[] ks = solution.getBaseSolutions();
                if (ks != null) {
                    base = Vector.addAll(base, ks);
                }
                Matrix<Double> re1 = Matrix.multiply(matrix.subMatrix(0, 0, row - 1, column - 2), base),
                        re2 = matrix.subMatrix(0, column - 1, row - 1, column - 1);
                assertTrue(re1.valueEquals(re2));
            }
		}
	}

//	@Test
	public void testSolveHomoEquation() {
		int row = 8;
		int column = 11;
		Vector<Double> zero = Vector.zeroVector(row, mcd);
		for (int i = 0; i < 100; i++) {
            double[][] mat = new double[row][];
            for (int j = 0; j < row; j++) {
                mat[j] = ArraySup.ranDoubleArr(column);
            }
            Matrix<Double> matrix = Matrix.of(mat).mapTo(x -> x, mcd);
            LinearEquationSolution<Double> solution = MatrixSup.solveHomogeneousLinearEquation(matrix);
            if (solution.getSolutionSituation() != LinearEquationSolution.Situation.NO_SOLUTION) {
                Vector<Double> base = solution.getSpecialSolution();
                Vector<Double>[] ks = solution.getBaseSolutions();
                if (ks != null) {
                    base = Vector.addAll(base, ks);
                }
                Matrix<Double> re = Matrix.multiply(matrix, base);
                assertTrue(re.valueEquals(zero));
            }
		}
	}
	
	@Test
	public void testEigenVector() {
        Matrix<Double> mat = Matrix.of(new double[][]{
                {0, 1, 1},
                {1, 0, 1},
                {1, 1, 0}
        });
        List<Pair<Double, Vector<Double>>> list = mat.eigenvaluesAndVectors(x -> {
            return Arrays.asList(-1d, -1d, 2d);
        });
        assertEquals("", list.toString(), "[[-1.0,[1.0,-1.0,0.0]], [-1.0,[1.0,0.0,-1.0]], [2.0,[-1.0,-1.0,-1.0]]]");
    }
	@Test
	public void testNormalizeUsingMatrix() {
		ConicSection<Double> cs = GeneralConicSection.generalFormula(
				2d, -Math.sqrt(3),1d,
				0d,0d,-10d, mcd);
		ConicSection<Double> afterTrans = cs.toStandardForm();
		assertTrue("",mcd.isEqual(afterTrans.getA(),2.5)&&mcd.isEqual(afterTrans.getC(),0.5));
		assertEquals("",cs.determineType(),ConicSection.Type.ELLIPSE);
	}

	@Test
    public void testDeterminant() {
        Matrix<Integer> mat = Matrix.of(new int[][]{
                {1, 2, 3, 4},
                {0, 3, 4, 5},
                {0, 0, 5, 6},
                {0, 0, 0, 4}
        });
        assertEquals(mat.calDet().intValue(), 3 * 5 * 4);
        var mat2 = Matrix.of(new double[][]{
                {1, 2, 3, 4, 5},
                {8, 0, 3, 4, 5},
                {27, 1.3, 5, 5, 6},
                {3, 4, 6, 7, 4},
                {1, 11, 3, -4, 3}
        });
        assertTrue(Math.abs(mat2.calDet() - MatrixSup.fastDet(mat2)) < 0.00001);
    }

    @Test
    public void testDecompositionQR() {
        Matrix<Fraction> A = Matrix.of(new int[][]{
                {1, 2, 3, 4},
                {2, 3, 4, 5},
                {3, 2, 5, 6},
                {6, -2, -8, 4}
        }).mapTo(Fraction::of, Fraction.getCalculator());
        var t = A.decompLU();
        var P = t.getFirst();
        var L = t.getSecond();
        var U = t.getThird();
//        P.printMatrix();
//        L.printMatrix();
        assertTrue("L should be lower triangular.", isUpperTriangular(L.transpose()));
        assertTrue("U should be upper triangular.", isUpperTriangular(U));
//        U.printMatrix();
        var m1 = Matrix.multiply(P, A);
        var m2 = Matrix.multiply(L, U);
        assertTrue("PA = LU", m1.valueEquals(m2));
    }
}

File: test/test/math/TestMatrix.kt
/**
 * 2018-01-25
 */
package test.math

import cn.ancono.math.MathCalculator
import cn.ancono.math.algebra.linear.LinearEquationSolution
import cn.ancono.math.algebra.linear.Matrix
import cn.ancono.math.algebra.linear.Matrix.Companion.of
import cn.ancono.math.algebra.linear.Vector
import cn.ancono.math.equation.SVPEquation
import cn.ancono.math.geometry.analytic.plane.curve.ConicSection
import cn.ancono.math.geometry.analytic.plane.curve.GeneralConicSection
import cn.ancono.math.numberModels.Calculators
import cn.ancono.math.numberModels.Fraction
import cn.ancono.math.numberModels.Fraction.Companion.calculator
import cn.ancono.math.numberModels.api.minus
import cn.ancono.math.numberModels.api.plus
import cn.ancono.math.numberModels.api.times
import org.junit.Assert
import org.junit.Test
import java.util.*
import kotlin.test.assertEquals

/**
 * @author liyicheng 2018-01-25 17:38
 */
class TestMatrix
/**
 *
 */
{
    var mc: MathCalculator<Long> = Calculators.longCal()
    var mcd = Calculators.doubleDev()

    private fun <T> isUpperTriangular(m: Matrix<T>): Boolean {
        val mc = m.mathCalculator
        val n = Math.min(m.row, m.column)
        for (i in 0 until n) {
            for (j in 0 until i) {
                if (!mc.isZero(m[i, j])) {
                    return false
                }
            }
        }
        return true
    }

    @Test
    fun testMultiply() {
        val A = of(2, 3, mcd) { i, j ->
            i + j + 0.0
        }
        val B = of(3, 3, mcd) { i, j ->
            i * j + 1.0
        }
//        println(A)
//        println(B)
        val C = of(2, 3, mcd,
                3.0, 8.0, 13.0,
                6.0, 14.0, 22.0)
        TestUtils.assertValueEquals(A * B, C)
    }

    @Test
    fun testAdd() {
        val A = of(2, 3, mcd) { i, j ->
            i + j + 0.0
        }
        val B = of(2, 3, mcd) { i, j ->
            i - j + 0.0
        }
        val C = of(2, 3, mcd) { i, j ->
            2.0 * i
        }
        TestUtils.assertValueEquals(C, A + B)
        assert((A - A).isZero())
    }

    @Test
    fun testToTriangle() {
        val A = of(2, 3, mcd) { i, j ->
            i + j + 0.0
        }
        val (r, ops) = A.toUpperTriangleWay()
        println(r)
        println(A.toEchelonWay().first)
//        println(A.toUpperTriangle())

    }

    @Test
    fun testSolve1() {
        val A = of(2, 3, mcd) { i, j ->
            i + j + 1.0
        }
        val solution = Matrix.solveLinearExpanded(A)
        assertEquals(LinearEquationSolution.SolutionType.SINGLE, solution.type)
        val special = Vector.of(mcd, -1.0, 2.0)
        TestUtils.assertValueEquals(special, solution.special)

    }

    @Test
    fun testSolve2() {
        val A = of(2, 3, mcd,
                1.0, 1.0, 2.0,
                2.0, 2.0, 3.0)
        val solution = Matrix.solveLinearExpanded(A)
        assertEquals(1, solution.solutionSpace.rank)
        assertEquals(LinearEquationSolution.SolutionType.EMPTY, solution.type)
    }

    @Test
    fun testSolve3() {
        var A = of(3, 2, mcd,
                1.0, 1.0,
                2.0, 2.0,
                1.0, 2.0
        )
        var basis = Matrix.solveHomo(A)
        assertEquals(0, basis.rank)

        A = of(
                2, 2, mcd,
                1.0, 1.0,
                2.0, 2.0,
        )
        basis = Matrix.solveHomo(A)
        assertEquals(1, basis.rank)


    }

    @Test
    fun testSolve4() {
        val A = of(3, 4, mcd,
                1.0, 2.0, 3.0, 1.0,
                1.0, 2.0, 1.0, -2.0,
                0.0, 0.0, 2.0, 4.0
        )
        val nullSpace = A.nullSpace()
        val N = Matrix.fromVectors(nullSpace.vectors)
        assert((A * N).isZero())
    }


    @Test
    fun testEigenEquation() {
        var mat = of(arrayOf(arrayOf(1L, 0L), arrayOf(0L, 4L)), Calculators.longCal())
        var equation = mat.charEquation()
        var expected: SVPEquation<Long> = SVPEquation.quadratic(1L, -5L, 4L, mc)
        Assert.assertTrue("EigenEquation:", expected.valueEquals(equation))
        mat = of(arrayOf(arrayOf(1, 2), arrayOf(3, 4)), Calculators.longCal())
        equation = mat.charEquation()
        expected = SVPEquation.quadratic(1L, -5L, -2L, mc)
        Assert.assertTrue("EigenEquation:", expected.valueEquals(equation))
        mat = of(arrayOf(arrayOf(1, 2, 3), arrayOf(3, 4, 5), arrayOf(4, 5, 6)), Calculators.longCal())
        equation = mat.charEquation()
        expected = SVPEquation.valueOf(mc, 0L, -9L, -11L, 1L)
        Assert.assertTrue("EigenEquation:", expected.valueEquals(equation))
    }

    @Test
    fun testSolveEquation() {
        val rd = Random()
        for (i in 0..99) {
            val row = rd.nextInt(10) + 1
            val column = rd.nextInt(10) + 1
            val expanded = of(row, column + 1, mcd) { _, _ ->
                rd.nextDouble()
            }
            val m = expanded.subMatrix(0, 0, row, column)
            val b = expanded.getColumn(column)
            val solution: LinearEquationSolution<Double> = Matrix.solveLinearExpanded(expanded)
            if (solution.notEmpty()) {
                val base = solution.special
                val t = Vector.multiplyToVector(m, base)
                TestUtils.assertValueEquals(b, t)
            }
        }
    }

    @Test
    fun testSolveHomoEquation() {
        val rd = Random()
        for (i in 0..99) {
            val row = rd.nextInt(10) + 1
            val column = rd.nextInt(10) + 1
            val matrix: Matrix<Double> = of(row, column, mcd) { _, _ ->
                rd.nextDouble()
            }
            val solution = Matrix.solveHomo(matrix)
            if (solution.rank > 0) {
                val n = Matrix.fromVectors(solution.vectors)
                val re = matrix * n
                Assert.assertTrue(re.isZero())
            }

//            if (!re.isZero()) {
//                println(re)
//                println(matrix)
//                println(matrix.toEchelonWay().first)
//                println(solution)
//            }


        }
    }

    @Test
    fun testEigenVector() {
        val mat: Matrix<Double> = of(arrayOf(arrayOf(0.0, 1.0, 1.0), arrayOf(1.0, 0.0, 1.0), arrayOf(1.0, 1.0, 0.0)), mcd)
        val list = mat.eigenvaluesAndVectors { listOf(-1.0, -1.0, 2.0) }
//        println(list.toString())
        assertEquals(3, list.size)
//        Assert.assertEquals("", list.toString(), "[[-1.0,[1.0,-1.0,0.0]], [-1.0,[1.0,0.0,-1.0]], [2.0,[-1.0,-1.0,-1.0]]]")
    }

    @Test
    fun testNormalizeUsingMatrix() {
        val cs: ConicSection<Double> = GeneralConicSection.generalFormula(
                2.0, -Math.sqrt(3.0), 1.0,
                0.0, 0.0, -10.0, mcd)
        val afterTrans = cs.toStandardForm()
        Assert.assertTrue("", mcd.isEqual(afterTrans.a, 2.5) && mcd.isEqual(afterTrans.c, 0.5))
        Assert.assertEquals("", cs.determineType(), ConicSection.Type.ELLIPSE)
    }

    //    @Test
    //    public void testDeterminant() {
    //        Matrix<Integer> mat = Matrix.of(new int[][]{
    //                {1, 2, 3, 4},
    //                {0, 3, 4, 5},
    //                {0, 0, 5, 6},
    //                {0, 0, 0, 4}
    //        });
    //        assertEquals(mat.calDet().intValue(), 3 * 5 * 4);
    //        var mat2 = Matrix.of(new double[][]{
    //                {1, 2, 3, 4, 5},
    //                {8, 0, 3, 4, 5},
    //                {27, 1.3, 5, 5, 6},
    //                {3, 4, 6, 7, 4},
    //                {1, 11, 3, -4, 3}
    //        });
    //        assertTrue(Math.abs(mat2.calDet() - MatrixSup.fastDet(mat2)) < 0.00001);
    //    }
    @Test
    fun testDecompositionLU() {
        val A: Matrix<Fraction> = of(arrayOf(
                arrayOf(1, 2, 3, 4),
                arrayOf(2, 3, 4, 5),
                arrayOf(3, 2, 5, 6),
                arrayOf(6, -2, -8, 4)), Calculators.integer())
                .mapTo(calculator) { Fraction.of(it.toLong()) }
        val (P, L, U) = A.decompLU()
        //        P.printMatrix();
//        L.printMatrix();
        Assert.assertTrue("L should be lower triangular.", isUpperTriangular(L.transpose()))
        Assert.assertTrue("U should be upper triangular.", isUpperTriangular(U))
        //        U.printMatrix();
        val m1 = P * A
        val m2 = L * U
//        println(P * A)
//        println(L * U)
        Assert.assertTrue("PA = LU", m1.valueEquals(m2))
    }

    @Test
    fun testDecompCholesky() {
        val rd = Random()
        val B: Matrix<Double> = of(5, 5, Calculators.doubleDev()) { _: Int?, _: Int? -> rd.nextDouble() }
        val A = B.multiply(B.transpose())
        val L = A.decompCholesky()
        //        A.congruenceDiagForm().getFirst().printMatrix();
//        L.printMatrix();
//        A.printMatrix();
        val R = L * L.transpose()
        //        R.printMatrix();
//        var A = Matrix.of(new double[][]{
//                {4, -1, 1},
//                {-1, 4.25, 2.75},
//                {1, 2.75, 3.5}
//        });
//        var L = A.decompCholesky();
//        L.printMatrix();
//        var R = Matrix.multiply(L,L.transpose());
        Assert.assertTrue("A = LL^T", A.valueEquals(R))
    }

    @Test
    fun testDecompCholeskyD() {
        val rd = Random()
        val B: Matrix<Double> = of(5, 5, Calculators.doubleDev()) { _: Int?, _: Int? -> rd.nextDouble() }
        val A = B * B.transpose()
        val (L, D) = A.decompCholeskyD()

//        A.congruenceDiagForm().getFirst().printMatrix();
//        L.printMatrix();
//        A.printMatrix();
        val R = Matrix.product(L, Matrix.diag(D), L.transpose())

//        R.printMatrix();
//        var A = Matrix.of(new double[][]{
//                {4, -1, 1},
//                {-1, 4.25, 2.75},
//                {1, 2.75, 3.5}
//        });
//        var L = A.decompCholesky();
//        L.printMatrix();
//        var R = Matrix.multiply(L,L.transpose());
        Assert.assertTrue("A = LDL^T", A.valueEquals(R))
    }
}